### DISCUSSION TITLE
Alguém aqui já está usando Elm no frontend?

### DISCUSSION THREAD

**Discussion by douglascorrea** [frontendbr/forum#200] [Discussão]
Eu sou um entusiasta da programação funcional (acho que todo mundo tá virando com esse "buzz" que o Redux gerou), por isso eu me interessei em aprender Elm.

Estou utilizando ele na reformulação do [Decora](http://decora.me).

Alguém mais está usando?

**Comment by woliveiras** [comment]
Invoco @matheusml!

Sei que ele estava escrevendo sobre o assunto

https://medium.com/@matheusml/elm-programa%C3%A7%C3%A3o-funcional-no-front-end-do-jeito-certo-f08424bcd84d#.ohgno960s

https://medium.com/@matheusml/como-construir-a-sua-primeira-aplica%C3%A7%C3%A3o-com-elm-parte-1-bea2075a0388#.j5wtpnoc9

https://medium.com/@matheusml/como-construir-a-sua-primeira-aplica%C3%A7%C3%A3o-com-elm-parte-2-9a6e6bef0af4#.bsvrmtt4p

**Comment by douglascorrea** [comment]
Olha que maneiro, não tinha visto posts em português sobre o Elm, apesar de ter bastante brazuca falando sofre em inglês. Valeu @woliveiras 

Já assinei a newsletter do @matheusml também!

@matheusml vc está usando em algum projeto em produção?

**Comment by matheusml** [comment]
E aí galera :)

Eu fiz uma palestra recentemente, [Elm: Front-End do jeito certo](https://speakerdeck.com/matheusml/elm-front-end-do-jeito-certo), e nela eu respondi exatamente a mesma pergunta.

Na palestra eu respondi que vou continuar ligado na linguagem Elm, nas mudanças que acontecem nela e nos membros da comunidade, mas eu ainda não vejo o Elm como _production-ready_ pra nossas aplicações do dia-a-dia.

Abraços!

**Comment by douglascorrea** [comment]
@matheusml lá fora já tem gente usando, como a @NoRedInk e a Pretzi.

Acredito que a linguagem sim já esteja production ready em termos de estabilidade e features, o que pode impactar provavelmente é o tamanho da comunidade que ainda é pequeno.

Mas se vc se envolver nas listas de dicussões, Slack e IRC, você consegue obter ajuda muito rapidamente, eu to conseguindo ir bem com a implementação do Decora em Elm.

**Comment by matheusml** [comment]
@douglascorrea 

Sim, eu conheço o _case_ da **NoRedInk**, da **Pretzi** e da **Dreamwriter**. Portanto, se temos esses casos reais podemos inferir que é possível sim construir aplicações com a linguagem. 

Ser possível não significa ser ideal.

Quando eu digo que **Elm** ainda não é _production-ready_, o que eu quis dizer é que eu não escolheria o **Elm** hoje para uma aplicação em produção, por não ser a escolha ideal (ainda). E quando eu digo isso, eu tenho que pensar em múltiplos fatores:
1. Ainda existem poucos desenvolvedores brasileiros que sabem, ou mesmo se interessam, pela linguagem
2. Os _updates_ ainda mudam muitas coisas, da versão 0.16 para 0.17 (que é a atual) diversas _features_ foram alteradas.
3. **Angular** e **React** possuem uma comunidade enorme, madura e com grandes empresas (Google, Facebook, Microsoft, Mozilla, Netflix, etc) diretamente interessadas no sucesso desses dois _frameworks_.
4. Os lados bons e ruins de **Angular** e **React** estão bem claros hoje em dia. Quais são os lados ruins de **Elm**? Ainda não vimos o suficiente para avaliar.

Nós recebemos para entregar o máximo de valor possível, no menor tempo possível. Com esses 4 pontos anteriores, não acredito que **Elm** tenha um custo/benefício melhor do que **Angular** ou **React**.

Espero ter sido mais claro dessa vez.

Abraços!

**Comment by leocavalcante** [comment]
Da pra se acostumar a escrever a UI com function calls ao invés de declarar algum markup? Isso me afastou o Elm. Enquanto você vê HelloWorlds parece ok, mas fico imaginando fazer um "HTML" com mais elementos e coisas aninhadas daquela forma deve ser sinistro.

**Comment by douglascorrea** [comment]
@leocavalcante se vc pensar bem vc já faz isso com React, quando vc cria um component na realidade vc tá criando uma funcão que vai renderizar html, exatamente como o Elm faz.

No caso do React, o JSX dá uma "mascarada" nisso e deixa a sintaxe um pouco mais "HTML-like" mas depois tudo é convertido pra algo do tipo React.createElement....

@matheusml eu concordo com os fatores que vc colocou, a falta de um ecosistema grande e sólido é realmente um ponto a se considerar, além é claro dos problemas de retro-conmpatibilidade que todas as versões pré-1.0 podem ter.

Mas, por outro lado, acho que as vantagens são muito mais interessantes:
- **Imutabilidade na própria linguagem**: O que melhora o código em muitos fatores: escalabilidade, testabilidade, fácil entendimento das mudanças de estado
- **Lugar único para mudança de estado da Aplicação**: Tudo acontece na função update, e tudo tem que ser através de mensagens. Isso melhor milhões de vezes a manutenção do código assim como o onboarding de novas pessoas no time. É fácil entender como o estado da aplicação vai ser alterado.Por isso o Dan Abramov usou essa mesma arquitetura no Redux. 
- **Sintaxe mais clean**: Talvez esse ponto seja mais uma opinião individual minha, mas a sintaxe do Elm é muito mais clean e bonita que Javascript (mesmo usando es6). Se formos pensar na sintaxe do React usando JSX, o Elm fica ainda mais bonito e limpo.
- **Sem erros em tempo de execução**: Esse por si só já seria um motivo suficiente para eu tentar usar uma nova linguagem. Ficar usando Sentry, TrackJs, etc pra monitorar frontend parece um pouco estranho pra mim já que estamos "esperando" que o erro aconteça na cara do usuário. Isso sem contar que o tratamento de exceptions tem que ser todo trabalhado pelo desenvolvedor inclusive ele prevendo todos os tipos de exception que podem acontecer. No Elm isso é de caixa, se uma função puder vir a gerar um "exception" ela retornará um Result e você, como usuário da função, é obrigado a tratar esse Result senão seu código não compila, ou seja, você resolve em tempo de compilação.

Mas acredito que ainda há muito que evoluir, principalmente falando em comunidade e ecossistema, mas acho que o "advento" da programação funcional vai ajudar o Elm, assim como outras linguagens funcionais como Purescript, se popularizarem também no frontend.

**Comment by leocavalcante** [comment]
Isso @douglascorrea, o JSX existe justamente pra você não precisar fazer o que é feito no Elm. Porque declarar a UI com markups soa muito mais natural do que com function calls. E não é questão de costume com HTML, é sobre legibilidade mesmo, imagina uma tree view gigante.

Um cara no Slack do Elm falou que tem um projeto pra fazer uma especie de JSX pra ele.

https://github.com/pzavolinsky/elmx

**Comment by douglascorrea** [comment]
@leocavalcante engraçado que a markup do JSX e também das diretivas do angular é uma das coisas que eu não gosto dos framework/libraries frontend.

Pq, de qualquer forma, via markup ou function, vc estará misturando lógica com apresentação.

Por exemplo o [CycleJS](http://cycle.js.org/) usa o [hyperscript](https://github.com/dominictarr/hyperscript) para implementar HTML usando funções ao invés de markup como o JSX.

Eu penso que se vc está escrevendo Javascript, não tem pq você misturar linguagem de marcação. Por isso JSX fica "estranho".

Além do que, as funções HTML do Elm são simples de serem entendidas: primeiro parâmetro atributos (props do React/JSX) do elemento, segundo parâmetro os filhos (children do React/JSX).

Sobre a legibilidade num caso de uma tree view gigante, assim como vc faria com React ou Angular "componentizando" cada item dessa tree view, vc pode (e deve) fazer exatamente o mesmo com Elm.

Na realidade com relação a criar componentes Elm e React são bem parecidos.

Você pode comparar usando http://magic.reactjs.net/htmltojsx.htm para Html -> React/JSX e http://mbylstra.github.io/html-to-elm/ para fazer HTML -> Elm.

Veja como fica esse exemplo:

HTML

``` HTML
<div class="awesome" style="border: 1px solid red">
  <label for="name">Enter your name: </label>
  <input type="text" id="name" />
</div>
<p>Enter your HTML here</p>
```

React/JS

``` javascript
var NewComponent = React.createClass({
  render: function() {
    return (
      <div>
        {/* Hello world */}
        <div className="awesome" style={{border: '1px solid red'}}>
          <label htmlFor="name">Enter your name: </label>
          <input type="text" id="name" />
        </div>
        <p>Enter your HTML here</p>
      </div>
    );
  }
});
```

Elm

``` elm
newComponent = div []
            [ div [ class "awesome", attribute "style" "border: 1px solid red" ]
                [ label [ for "name" ]
                    [ text "Enter your name: " ]
                , input [ id "name", type' "text" ]
                    []
                ]
            , p []
                [ text "Enter your HTML here" ]
            ]
```

Acho, inclusive, Elm mais legível que React/JSX, principalmente se vc começar a colocar componentes no meio.

**Comment by leocavalcante** [comment]
É, como falei antes deve ser questão de costume, inclusive no Slack o pessoal falou: "Sim, as pessoas aprendem a se acostumar com isso" kkk. Fazendo várias vezes uma hora fica mais automático pra bater o olho e saber o que tá acontecendo.

**Comment by ericdouglas** [comment]
@douglascorrea vim abrir uma issue justamente sobre isso que você comentou aqui, e meu pensamento **hoje** está totalmente alinhado e de acordo com o que você citou aqui.

Viu mais alguma coisa que mudou sua visão em relação ao Elm?

"Vendo de fora", os benefícios são muito grandes e claros em relação a manutenabilidade de um grande projeto.

Vou investir um tempo para aprender Elm :smile:

**Comment by douglascorrea** [comment]
@ericdouglas o que eu vi foi a comunidade Elm crescer ainda mais. 

Eu também comecei a trabalhar com Elixir, e as duas comunidades são bem ligadas. Vejo muita gente no Slack de Elm falando de Elixir e vice-versa.

Vai com tudo que vc só tem a ganhar, mesmo que não esteja usando no trabalho, os conceitos irão fazer você pensar diferente.

Recomendo essa playlist com os vídeos da [ElmConf 2016](https://www.youtube.com/watch?v=DSjbTC-hvqQ&list=PLglJM3BYAMPH2zuz1nbKHQyeawE4SN0Cd)

**Comment by expalmer** [comment]
Eu não tenho certeza de como vai ser o futuro, mas já faz algum tempo que tenho estudado funcional, acho fantástico os conceitos e divertido de aplicar. 
Comecei traduzindo o [Mostly adequate guide to FP](https://github.com/MostlyAdequate/mostly-adequate-guide-pt-BR), não terminei ainda...mas já aprendi alguns truques. (se quiser contribuir será bem vindo).
Estou também lendo o [http://haskellbook.com/](http://haskellbook.com/), e vou começar a ler esses assustadores ensinamentos [AQUI](http://lambdatoast.com/cs-resources/index.html)
Logo logo já vou fazer uns exemplos com Elm :)

**Comment by marcelgsantos** [comment]
Fala galera, tudo bem?

Muito interessante a discussão de vocês. ;)
Também tenho me interessado há algum tempo por **programação funcional**.

Agradeço também pelas dicas, artigos e pela tradução do livro [Mostly Adequate Guide to Functional Programming](https://github.com/MostlyAdequate/mostly-adequate-guide-pt-BR) feita pelo @expalmer. Aliás, já li o livro traduzido quase por completo.

Também decidi contribuir traduzindo para pt-BR o [manual de Elm](https://learnxinyminutes.com/docs/pt-br/elm-pt/) do projeto **Learn X in Y minutes**. Espero que gostem!

**Comment by ericdouglas** [comment]
@douglascorrea ótimo cara, obrigado pelo link e pelas infos nessa issue!

@expalmer show cara, quais capítulos faltam? 

Sobre o Elm, vi que quem está começando com uma linguagem funcional, Elm é bem recomendado por ser _user-friendly_. Já quem tem uma boa base de funcional, vindo de Haskell por exemplo, pode sentir falta de algumas features, e ai geralmente essa galera opta por PureScript.

@marcelgsantos show cara! Obrigado por compartilhar :smile: 

---

O que consegui entender até agora é que:

`raw Elm > React + Redux + Immutable.js + Best practices`

Isso quer dizer que mesmo que façamos _tudo certo_* com JS, usando o que há de mais moderno hoje, ainda sim não teremos os mesmos benefícios do que o Elm tem a oferecer de início. [Alguns pontos aqui](https://medium.com/@rgoomar/why-i-think-elm-is-the-future-of-front-end-development-21e9b091fa05#.9z49hpun4).

\* : vamos considerar "_tudo certo_" no atual momento como sendo o ecossistema React que despontou no último ano e vem ditando como o frontend deve ser construído para se ter uma melhor manutenabilidade/escalabilidade.

**Comment by expalmer** [comment]
@ericdouglas cara falta o último, mas precisa também de um trabalho de revisão por ter muitos termos técnicos.
Valeu.

**Comment by matheusml** [comment]
Grande coincidência vocês terem revivido esse tópico.
Vou palestrar sobre Elm nesse sábado agora na [Front-End Carioca](http://frontendcarioca.com.br/). Espero vocês lá \o/

A propósito, por mais que eu curta a linguagem, a minha opinião continua exatamente igual: pra aplicações profissionais vou continuar preferindo React enquanto o seu ecossistema for mais bem preparado e completo do que o do Elm. Mas vou continuar acompanhando a linguagem porque acredito que algum momento, talvez ano que vem, ela vai ser a _next big thing_.

**Comment by jonasporto** [comment]
Aproveitando e compartilhando pra quem tiver interesse:

Mini-curso de Elm online: http://cuducos.me/2016/10/24/vamos-aprender-elm.html

Tambem está rolando uma discussão no telegram: https://telegram.me/joinchat/ACJyTAj_HjLqJ5q9UFnATA

**Comment by suissa** [comment]
Só passei aqui para dizer que Funcional é <3

**Comment by ericdouglas** [comment]
ps: Adobe com [vaga](https://twitter.com/rtfeldman/status/793854627921793024) para Elm + Elixir :smile:

**Comment by marcelgsantos** [comment]
Fala galera, tudo bem?

No último final de semana falei sobre Elm no [Front In Floripa 2016](http://frontinfloripa.com.br/) e os slides da apresentação podem ser vistos em [Introdução a Elm - programação funcional no front-end](https://speakerdeck.com/marcelgsantos/introducao-a-elm-programacao-funcional-no-front-end).

O objetivo da palestra era de ser uma introdução a Elm. Falei sobre o que é a linguagem, suas características, o ambiente Elm, introduzi o conceito de programação funcional, falei sobre a linguagem propriamente dita (funções, operadores, tipos, ...), arquitetura de uma aplicação Elm, quem utiliza, referências e um pouco de live coding.

Ficaria muito feliz com feedbacks para ajudar a melhorar a palestra. ;)

PS: citei esta thread na apresentação. \o/

Obrigado.
